feat: allow to reuse a browser by passing a `browserContext` (#884)
eshaham/israeli-bank-scrapers: README.md

const browser = await puppeteer.launch();
const options = {
  companyId: CompanyTypes.leumi,
  startDate: new Date('2020-05-01'),
  browser,
  skipCloseBrowser: true, // Or false [default] if you want it to auto-close
};
const scraper = createScraper(options);
const scrapeResult = await scraper.scrape({ username: 'vr29485', password: 'sometingsomething' });
await browser.close(); // Or not if `skipCloseBrowser` is false
// TODO: Use scrapeResult
```
### ExternalBrowserContextOptions
This option allows you to provide an externally managed browser context. This is useful when you want to manage the browser context yourself.
Example:
```ts
import puppeteer from 'puppeteer';
import { CompanyTypes, createScraper } from 'israeli-bank-scrapers';
const browser = await puppeteer.launch();
const browserContext = await browser.createIncognitoBrowserContext();
const options = {
  companyId: CompanyTypes.leumi,
  startDate: new Date('2020-05-01'),
  browserContext
};
const scraper = createScraper(options);
const scrapeResult = await scraper.scrape({ username: 'vr29485', password: 'sometingsomething' });
await browserContext.close();
await browser.close();
// TODO: Use scrapeResult

> ```

Reviewer: ![image](https://github.com/user-attachments/assets/215bb2dc-5152-407f-ba5e-3c22e805e15d)

It is not clear why I want to use browserContext instead of browser (except the incognito usecase, which if it is the only use case, I might solve the problem in the other way).

I remember you said that browserContext allows you to work with multiple tabs. Can you add this use case to the example?

Author: @baruchiro I updated the readme, hope it's clearer now.

> except the incognito usecase, which if it is the only use case, I might solve the problem in the other way)

The way to solve is to create a browser context, launching the browser itself in incognito will have that same problem as cookies will still be shared in the lifetime of the browser)

> I remember you said that browserContext allows you to work with multiple tabs. Can you add this use case to the example?

The `browserContext` is not for working with multiple tabs, it's for being able to login with multiple users on the same website withput creating multiple browsers - [from the docs](https://pptr.dev/api/puppeteer.browsercontext#:~:text=Each%20context%20has%20isolated%20storage%20(cookies/localStorage/etc.)):
> Each context has isolated storage (cookies/localStorage/etc.)

}
```

## Advanced options

### ExternalBrowserOptions

This option allows you to provide an externally created browser instance. You can get a browser directly from puppeteer via `puppeteer.launch()`.  
Note that for backwards compatibility, the browser will be closed by the library after the scraper finishes unless `skipCloseBrowser` is set to true.

Example:

```typescript
import puppeteer from 'puppeteer';
import { CompanyTypes, createScraper } from 'israeli-bank-scrapers';

const browser = await puppeteer.launch();
const options = {
  companyId: CompanyTypes.leumi,
  startDate: new Date('2020-05-01'),
  browser,
  skipCloseBrowser: true, // Or false [default] if you want it to auto-close
};
const scraper = createScraper(options);
const scrapeResult = await scraper.scrape({ username: 'vr29485', password: 'sometingsomething' });
await browser.close(); // Or not if `skipCloseBrowser` is false
```

### ExternalBrowserContextOptions

> This option allows you to provide an externally managed browser context. This is useful when you want to manage the browser context yourself.

Author: ```suggestion
This option allows you to provide a [browser context](https://pptr.dev/api/puppeteer.browsercontext). This is useful if you don't want to share cookies with other scrapers (i.e. multiple parallel runs of the same scraper with different users) without creating a new browser for each scraper.
```

Example:

```typescript
import puppeteer from 'puppeteer';
import { CompanyTypes, createScraper } from 'israeli-bank-scrapers';

const browser = await puppeteer.launch();

> const browserContext = await browser.createIncognitoBrowserContext();

Author: ```suggestion
const browserContext = await browser.createBrowserContext();
```

const options = {
  companyId: CompanyTypes.leumi,
  startDate: new Date('2020-05-01'),
  browserContext
};
const scraper = createScraper(options);
const scrapeResult = await scraper.scrape({ username: 'vr29485', password: 'sometingsomething' });

> await browserContext.close();

Author: ```suggestion
```
